File: component-starter-server/src/main/frontend/src/App/Component/Schema.js
/**
 *  Copyright (C) 2006-2019 Talend Inc. - www.talend.com
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import React from 'react';
import classnames from 'classnames';
import Node from './Node';
import TileContext from '../tile';

import theme from './Schema.scss';
import SelectDataset from '../components/SelectDataset';

export default class Schema extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			schema: props.schema,
		};
		this.onSelectDataset = this.onSelectDataset.bind(this);
	}

	componentWillReceiveProps(nextProps) {
		if (this.props !== nextProps) {
			this.setState({
				schema: nextProps.schema,
			});
			!!this.props.onReload && this.props.onReload();
		}
	}

	onSelectDataset(service) {
		return (event, dataset) => {
			if (event.target.value==='create-new') {
				const tile = (
					<form className="form">
						<h2>Create a new dataset</h2>
						<div className="form-group">
							<label></label>
							<Schema schema={{entries: []}} readOnly={true} name="dataset" />
						</div>
					</form>
				);
				service.addTile(tile);
			} else {
				this.setState(prevState => {
					let replace = false;
					if (prevState.schema.entries.length > 0) {
						if (prevState.schema.entries[0].type === 'dataset') {
							replace = true;
							prevState.schema.entries[0].name = dataset;
						}
					}
					if (!replace) {
						prevState.schema.entries.push({ name: dataset, type: 'dataset' });
					}
					return Object.assign({}, prevState);
				});
			}
		}
	}

	render() {
		return (
			<form className={classnames('form', theme.Schema)}>
				{this.props.name === 'configuration' && (
					<TileContext.Consumer>
						{(service) => (
							<SelectDataset onChange={this.onSelectDataset(service)} />
						)}
					</TileContext.Consumer>
				)}
				<div className="form-group">
					<label htmlFor="schema-model">Model</label>
					<Node
						node={this.state.schema}
						readOnly={this.props.readOnly || !!this.props.parent}
						name={this.props.name}
					/>
				</div>
			</form>
		);
	}
}

File: component-server-parent/component-server-model/src/main/java/org/talend/sdk/component/server/front/model/Schema.java
/**
 * Copyright (C) 2006-2025 Talend Inc. - www.talend.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.talend.sdk.component.server.front.model;

import java.beans.ConstructorProperties;
import java.math.BigDecimal;
import java.time.temporal.Temporal;
import java.util.Collection;
import java.util.Date;
import java.util.List;
import java.util.Map;

import lombok.Data;

@Data
public final class Schema {

  private final Type type;

  private final Schema elementSchema;

  private final List<Entry> entries;

  private final List<Entry> metadata;

  private final Map<String, String> props;

  @ConstructorProperties({"type", "elementSchema", "entries", "metadata", "props"})
  public Schema(
          final Type type,
          final Schema elementSchema,
          final List<Entry> entries,
          final List<Entry> metadata,
          final Map<String, String> props) {
    this.type = type;
    this.elementSchema = elementSchema;
    this.entries = entries;
    this.metadata = metadata;
    this.props = props;
  }

  public String getProp(final String key){
      return this.props.get(key);
  }

  public enum Type {

    RECORD(new Class<?>[] { Record.class }),
    ARRAY(new Class<?>[] { Collection.class }),
    STRING(new Class<?>[] { String.class, Object.class }),
    BYTES(new Class<?>[] { byte[].class, Byte[].class }),
    INT(new Class<?>[] { Integer.class }),
    LONG(new Class<?>[] { Long.class }),
    FLOAT(new Class<?>[] { Float.class }),
    DOUBLE(new Class<?>[] { Double.class }),
    BOOLEAN(new Class<?>[] { Boolean.class }),
    DATETIME(new Class<?>[] { Long.class, Date.class, Temporal.class }),
    DECIMAL(new Class<?>[] { BigDecimal.class });

    /**
     * All compatibles Java classes
     */
    private final Class<?>[] classes;

    Type(final Class<?>[] classes) {
      this.classes = classes;
    }

    /**
     * Check if input can be affected to an entry of this type.
     *
     * @param input : object.
     *
     * @return true if input is null or ok.
     */
    public boolean isCompatible(final Object input) {
      if (input == null) {
        return true;
      }
      for (final Class<?> clazz : classes) {
        if (clazz.isInstance(input)) {
          return true;
        }
      }
      return false;
    }
  }
}